homeassistant/components/unifiprotect/binary_sensor.py

```diff
@@ -216,33 +226,43 @@ def _async_update_device_from_protect(self) -> None:
 class ProtectDiskBinarySensor(ProtectNVREntity, BinarySensorEntity):
     """A UniFi Protect NVR Disk Binary Sensor."""
 
+    _disk: UOSDisk
     entity_description: ProtectBinaryEntityDescription
 
     def __init__(
         self,
         data: ProtectData,
         device: NVR,
         description: ProtectBinaryEntityDescription,
-        index: int,
+        disk: UOSDisk,
     ) -> None:
         """Initialize the Binary Sensor."""
+        self._disk = disk
+        # backwards compat with old unique IDs
+        index = self._disk.slot - 1
+
         description = copy(description)
         description.key = f"{description.key}_{index}"
-        description.name = (description.name or "{index}").format(index=index)
-        self._index = index
+        description.name = f"{disk.type} {disk.slot}"
         super().__init__(data, device, description)
 
     @callback
     def _async_update_device_from_protect(self) -> None:
         super()._async_update_device_from_protect()
 
-        disks = self.device.system_info.storage.devices
-        disk_available = len(disks) > self._index
-        self._attr_available = self._attr_available and disk_available
-        if disk_available:
-            disk = disks[self._index]
-            self._attr_is_on = not disk.healthy
-            self._attr_extra_state_attributes = {ATTR_MODEL: disk.model}
+        slot = self._disk.slot
+        self._attr_available = False
+
+        if self.device.system_info.ustorage is None:
+            return
+
+        for disk in self.device.system_info.ustorage.disks:
+            if disk.slot == slot:
+                self._disk = disk
+                self._attr_available = True
+                break
+
+        self._attr_is_on = self._disk.state not in HEALTHY_DISK_STATES
```

Comment:
Even better if the lib had a `.is_healthy` so Home Assistant doesn't have to know about what that means